src/mds/nameserver2/clean_manager.cpp

```diff
@@ -155,13 +157,22 @@ void CleanDiscardSegmentTask::ScanAndExecTask() {
             continue;
         }
 
+        auto count = discardSegments.size();
+        if (count == 0) {
+            continue;
+        }
+
+        curve::common::CountDownEvent counter(count);
         for (const auto& kv : discardSegments) {
-            if (!cleanManager_->SubmitCleanDiscardSegmentJob(kv.first,
-                                                             kv.second)) {
+            if (!cleanManager_->SubmitCleanDiscardSegmentJob(
+                    kv.first, kv.second, &counter)) {
                 LOG(ERROR) << "SubmitCleanDiscardSegmentJob failed";
-                continue;
+                counter.Signal();
             }
         }
+
+        // wait all submitted jobs finish
```

Comment:
wait all jobs submitted?

Reply:
Not true, the logic here is loading persistent tasks and issuing asynchronous jobs for each task.
And after the job is completed, it deletes the corresponding persistent task.
So, if don't wait for all submitted jobs to finish, the next loop will load persistent tasks and issue jobs for those tasks, but there may exist the same jobs that do the same thing because the previous job hasn't finished and its corresponding persistent task remain undeleted.

Reply:
ok